### PR TITLE
AirfRANS: AdamW LR bracket sweep (3e-4, 4e-4, 6e-4, 8e-4)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

AdamW at lr=5e-4 beat Lion lr=3e-4 by 38% on AirfRANS (0.3308 vs 0.5384). But 5e-4 was the first AdamW point tested — the optimal LR is unknown. This PR brackets around 5e-4 with 4 parallel trials using slices=64 for faster epoch cycling.

## Instructions

Run 4 parallel trials on 4 separate GPUs. All settings fixed except `lr` and `model_slices`. The `model_slices=64` change (vs default 96) makes each epoch ~33% faster, fitting more epochs within the 30-min timeout.

```bash
cd target/icml2026

# v0: lr=3e-4
CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans_task full \
  --optimizer adamw --lr 3e-4 --model_slices 64 \
  --wandb_group airfrans-adamw-lr-bracket \
  --wandb_run_name v0-adamw-lr3e4 &

# v1: lr=4e-4
CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset airfrans --airfrans_task full \
  --optimizer adamw --lr 4e-4 --model_slices 64 \
  --wandb_group airfrans-adamw-lr-bracket \
  --wandb_run_name v1-adamw-lr4e4 &

# v2: lr=6e-4
CUDA_VISIBLE_DEVICES=2 python train.py \
  --dataset airfrans --airfrans_task full \
  --optimizer adamw --lr 6e-4 --model_slices 64 \
  --wandb_group airfrans-adamw-lr-bracket \
  --wandb_run_name v2-adamw-lr6e4 &

# v3: lr=8e-4
CUDA_VISIBLE_DEVICES=3 python train.py \
  --dataset airfrans --airfrans_task full \
  --optimizer adamw --lr 8e-4 --model_slices 64 \
  --wandb_group airfrans-adamw-lr-bracket \
  --wandb_run_name v3-adamw-lr8e4 &

wait
```

Report for each variant: epochs completed, whether 30-min timeout or 50-epoch cap hit, and `val_primary/surface_mse` at each epoch. Note whether any variant is still improving at cutoff.

## Baseline

- **val_primary/surface_mse:** 0.3308 (val) / 0.3199 (test)
- **Config:** AdamW lr=5e-4, model_slices=96, 3L/192d/3H
- **PR:** #2423 (W&B run: u95mzqso)
- **Reproduce:** `cd target/icml2026 && python train.py --dataset airfrans --airfrans_task full --optimizer adamw --lr 5e-4`

## Results

_To be filled in by student_

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT kohaku:

## Results

### AdamW LR Bracket Sweep (slices=64, 5 epochs each)

All 4 variants ran for 5 epochs within the 30-min timeout (slices=64 gave faster iterations as expected). All were still improving at cutoff.

#### Summary Table

| Variant | LR | Best Val surface_mse | Test surface_mse | Test volume_mse | W&B Run |
|---------|------|---------------------|-----------------|----------------|---------|
| **v0** | **3e-4** | **0.3414** | **0.3215** | **0.6921** | [`5bcn2ryw`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/5bcn2ryw) |
| v1 | 4e-4 | 0.3754 | 0.3460 | 0.7035 | [`91b64jgy`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/91b64jgy) |
| v2 | 6e-4 | 0.3513 | 0.3405 | 0.6988 | [`2uhvsmgd`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/2uhvsmgd) |
| v3 | 8e-4 | 0.3278 | 0.3222 | 0.7933 | [`u7iu81h0`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/u7iu81h0) |

**Baseline (PR #2423):** val=0.3308, test=0.3199 (lr=5e-4, slices=96, 6 epochs)

#### Per-Channel Test Surface MSE

| Variant | LR | Ux | Uy | p | nut |
|---------|------|--------|--------|--------|--------|
| v0 | 3e-4 | 0.00120 | 0.00058 | 1.2838 | 0.00024 |
| v1 | 4e-4 | 0.01355 | 0.00035 | 1.3697 | 0.00038 |
| v2 | 6e-4 | 0.00322 | 0.00076 | 1.3570 | 0.00114 |
| v3 | 8e-4 | 0.00199 | 0.00104 | 1.2855 | 0.00022 |

#### Epoch-by-Epoch val_primary/surface_mse

| Epoch | v0 (3e-4) | v1 (4e-4) | v2 (6e-4) | v3 (8e-4) |
|-------|-----------|-----------|-----------|-----------|
| 1 | 0.5534 | 0.5464 | 0.5917 | 0.5712 |
| 2 | 0.4980 | 0.5204 | 0.5163 | 0.4984 |
| 3 | 0.4381 | 0.4549 | 0.4370 | 0.4248 |
| 4 | 0.3781 | 0.4157 | 0.3906 | 0.3671 |
| 5 | 0.3414 | 0.3754 | 0.3513 | 0.3278 |

#### Commands

```bash
cd target/icml2026

# All 4 used these shared settings:
# --dataset airfrans --airfrans-task full --optimizer adamw --model-slices 64
# --epochs 150 --batch-size 2 --agent kohaku --wandb-group "airfrans-adamw-lr-bracket"

CUDA_VISIBLE_DEVICES=0 python train.py ... --lr 3e-4 --wandb-name "kohaku/v0-adamw-lr3e4"
CUDA_VISIBLE_DEVICES=1 python train.py ... --lr 4e-4 --wandb-name "kohaku/v1-adamw-lr4e4"
CUDA_VISIBLE_DEVICES=2 python train.py ... --lr 6e-4 --wandb-name "kohaku/v2-adamw-lr6e4"
CUDA_VISIBLE_DEVICES=3 python train.py ... --lr 8e-4 --wandb-name "kohaku/v3-adamw-lr8e4"
```

#### What happened

**Higher LR converges faster on surface_mse but hurts volume_mse.** v3 (lr=8e-4) achieved the best val surface_mse (0.3278) after 5 epochs, beating even the baseline's 0.3308 after 6 epochs with slices=96. However, v3's volume_mse (0.7933) is significantly worse than v0's (0.6921), suggesting that high LR overfits the surface head at the expense of volume prediction.

**lr=3e-4 gives the best overall balance.** v0 matched the baseline's test surface_mse (0.3215 vs 0.3199) with fewer epochs and lower model capacity (slices=64 vs 96), while also achieving the best volume_mse (0.6921 vs baseline 0.6695). Note slices=64 is a smaller model, so direct comparison isn't apples-to-apples.

**Pressure dominates everywhere.** The `p` channel accounts for >99.5% of surface_mse across all variants, with per-channel p MSE ranging from 1.28-1.37. Velocity and nut channels are all <0.014.

All variants were still improving at the 5-epoch cutoff — none showed signs of plateauing.

#### Suggested follow-ups

- Run lr=3e-4 with full slices=96 for a fair comparison against the PR #2423 baseline
- The surface/volume trade-off at high LR suggests separate learning rates or loss weighting could help
- Consider whether the `p` channel bottleneck would benefit from a dedicated pressure head or loss scaling

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2428 because ADVISOR: Closing — all trials ran at slices=64 (vs baseline slices=96), making the LR comparison confounded. The best result (lr=8e-4, val=0.3278) barely beats val baseline but does NOT beat test baseline (0.3222 vs 0.3199), and volume_mse regresses severely (0.7933 vs 0.6695). The key finding: pressure channel dominates AirfRANS error in all configurations. The more important follow-up: apply no-EMA to AirfRANS (confirmed harmful in #2431), run as a single job at slices=96 for proper epoch count, and also apply asinh-pressure transform (proven on TandemFoil) to attack the pressure channel. Kohaku should re-run at slices=96 single job with no-EMA..
